### PR TITLE
nixos/k3s: add nftables to Path of k3s service

### PR DESCRIPTION
**Pull request by RMTT**
kubernetes 1.31 has nftables support, and k3s now support kubernetes 1.31, so  it needs nftables binary `nft`.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/cluster/k3s/default.nix
Comment (Mic92) [inline nixos/modules/services/cluster/k3s/default.nix:501]:
```suggestion
        path = [ pkgs.nftables ] ++ lib.optional config.boot.zfs.enabled config.boot.zfs.package;
```

File: nixos/modules/services/cluster/k3s/default.nix
Comment (superherointj) [inline nixos/modules/services/cluster/k3s/default.nix:501]: > ++ lib.optional config.networking.nftables.enable pkgs.nftables

How about only adding `nftable` in case `nftable` is enabled for host?

Comment (superherointj) [reply nixos/modules/services/cluster/k3s/default.nix:501]: If host doesn't have `nftables` enabled, how will `nftables` work?

> even i'm using iptables mode

In case, you have explicitly chosen `iptables mode` and K3s still requires `nftables` isn't that a bug somewhere else?

> so it is not detectable in nix configuration.

It should be. Same way there is a `extraKubeletConfig`, `extraKubeProxyConfig` is possible.
Likely would be convenient to have a Nix flag to toggle between iptables/nftables matching the host configuration by default.

**Comment by superherointj** [comment]
Please, to follow up with nixpkgs guidelines, properly scope the NixOS module name and rename title of this PR and commits. Such as:

> nixos/k3s: add nftables to path of k3s service

File: nixos/modules/services/cluster/k3s/default.nix
Comment (RMTT) [reply nixos/modules/services/cluster/k3s/default.nix:501]: > In case, you have explicitly chosen iptables mode and K3s still requires nftables isn't that a bug somewhere else?

Hmm, i searched the kubernetes code, and found it's just an intended behavior, and it's ok with this error. https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/cmd/kube-proxy/app/server_linux.go#L533

> It should be. Same way there is a extraKubeletConfig, extraKubeProxyConfig is possible.
Likely would be convenient to have a Nix flag to toggle between iptables/nftables matching the host configuration by default.

You're right, i will change code later.

Comment (RMTT) [reply nixos/modules/services/cluster/k3s/default.nix]: > Hmm, i searched the kubernetes code, and found it's just an intended behavior, and it's ok with this error. https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/cmd/kube-proxy/app/server_linux.go#L533

I still think the nftables should be added directly, not optional. For example, if i switch the kube-proxy mode from `nftable` to `iptables` or `ipvs` , the k3s also need the `nft` to clean the previous nftables rules.

**Comment by Mic92** [comment]
@superherointj happy with that one?

**Comment by superherointj** [comment]
> @superherointj happy with that one?

Yes.

**Comment by RMTT** [comment]
pls dont merge now, i'm testing it.

**Comment by Mic92** [comment]
Just undraft when ready.

**Comment by RMTT** [comment]
After searched kube-proxy code, i found the `kubeconfig` param will be override if we use `extraKubeProxyConfig`(https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/options.go#L221). That says, if user want to use `extraKubeProxyConfig`, the user must set `kubeconfig` (which it's a file generated by k3s):
```
  services.k3s = {
    enable = true;
    role = "server";
    extraKubeProxyConfig = {
      clientConnection = { kubeconfig = "/etc/rancher/k3s/k3s.yaml"; };
    };
  };
```
otherwise, the `kube-proxy` will throw an error to indicate it miss the kubernetes server and port configuration.

Do you think it's acceptable? Or just still use the `--kube-proxy-args` as before. @superherointj

File: nixos/modules/services/cluster/k3s/default.nix
Comment (superherointj) [reply nixos/modules/services/cluster/k3s/default.nix:523]: Okay, add `nftables` by default then.

**Comment by superherointj** [comment]
> if user want to use `extraKubeProxyConfig`, the user must set `kubeconfig` (which it's a file generated by k3s):
> 
>   `services.k3s.extraKubeProxyConfig.clientConnection.kubeconfig = "/etc/rancher/k3s/k3s.yaml";`
> 
> otherwise, the `kube-proxy` will throw an error to indicate it miss the kubernetes server and port configuration.
> 
> Do you think it's acceptable? Or just still use the `--kube-proxy-args` as before. @superherointj

Since that is required, `clientConnection.kubeconfig` can be a default value, that can be overridden.
We should **briefly** documented this behavior in `extraKubeProxyConfig` description or k3s docs in nixpkgs.

Unfortunately iptables/nftables switch adds another testing dimension. `iptables`, `nftables`, `kube-proxy-args` features needs tests or may break silently on upgrades since K3s maintainers trust tests for upgrades.

I get that this is extra work, so I'm not requiring you to do that. In case you prefer, just add the extra `nftables` package. Feel free to only go as far as you are comfortable with, and also feel free to stop with your initial suggestion that was just adding `nftables` package in case you prefer.

**Comment by superherointj** [comment]
All git commits should have a scoped title (`nixos/k3s: xxxx`). Since nixpkgs is a monorepo.

**Comment by RMTT** [comment]
> All git commits should have a scoped title (like, `nixos/k3s: xxxx`). Since nixpkgs is a monorepo.

i know, but i prefer to format commit message via squash merging

File: nixos/modules/services/cluster/k3s/default.nix
Comment (rorosen) [inline nixos/modules/services/cluster/k3s/default.nix:528]: Other packages that k3s needs are wrapped with the binaries. We should add `nftables` also there to be consistent with other dependencies.

See https://github.com/NixOS/nixpkgs/blob/1d7a088c34ef229519eafd6bee5d5d70a6b9df18/pkgs/applications/networking/cluster/k3s/builder.nix#L361

File: nixos/modules/services/cluster/k3s/default.nix
Comment (rorosen) [inline nixos/modules/services/cluster/k3s/default.nix:444]: I'm not sure whether I get your comments right, but I think this is a wrong default. K3s actually doesn't use this kubeconfig for kube-proxy. By default it passes kube-proxy the argument `--kubeconfig=/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig`, which is another file. I suggest to make the default `{}`, so that the `--kube-proxy-arg=config=...` gets only passed if a user explicitly changes this option.

**Review by rorosen — CHANGES_REQUESTED** [review]

File: nixos/modules/services/cluster/k3s/default.nix
Comment (RMTT) [reply nixos/modules/services/cluster/k3s/default.nix:444]: k3s will pass `kubeconfig` to kube-proxy in https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/daemons/agent/agent_linux.go#L43 via command line arg `--kube-config` , then kube-proxy will save this arg to `o.config.ClientConnection.Kubeconfig` in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/options.go#L106 . However, the `o.config` object will be override in https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-proxy/app/options.go#L221 by the kubeProxyConfigurataion object which specified via the `config=path` option.

Comment (rorosen) [reply nixos/modules/services/cluster/k3s/default.nix:444]: > k3s will pass kubeconfig to kube-proxy in https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/daemons/agent/agent_linux.go#L43 via command line arg --kube-config

I agree, but notice that the variable used here is `cfg.KubeConfigKubeProxy`, which gets constructed [here](https://github.com/k3s-io/k3s/blob/4ec261733e241e2baf77684ad5d87ff071d0f00a/pkg/agent/config/config.go#L518) . K3s generates multiple kubeconfigs with different certificates for different components, e.g. kubelet, kubeconfig, k3s-controller, etc. I don't know the internals but we shouldn't change which kubeconfig is used without good reason. The default kubeconfig for kube-proxy is at `/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig` and not the admin kubeconfig at `/etc/rancher/k3s/k3s.yaml`. 

You can see which args are passed to kube-proxy with `journalctl -u k3s | grep kube-proxy`, on the current master **without** this PR and default settings I see:

`k3s[778]: time="2024-12-08T14:05:20Z" level=info msg="Running kube-proxy --cluster-cidr=10.42.0.0/16 --conntrack-max-per-core=0 --conntrack-tcp-timeout-close-wait=0s --conntrack-tcp-timeout-established=0s --healthz-bind-address=127.0.0.1 --hostname-override=foo --kubeconfig=/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig --proxy-mode=iptables"`

The actual default for `extraKubeProxyConfig` would change the kubeconfig that kube-proxy uses. Of course we could change the default, but this would lead to other problems, for example if k3s changes its defaults or the k3s data dir is changed. I think it would be best to not generate a config for kube-proxy by default, so that k3s can decide which kubeconfig it passes to kube-proxy. However, if somebody configures a custom `extraKubeProxyConfig`, he has to take care of setting `clientConnection.kubeconfig` as mentioned in the option description.

Comment (RMTT) [reply nixos/modules/services/cluster/k3s/default.nix]: > The actual default for `extraKubeProxyConfig` would change the kubeconfig that kube-proxy uses. Of course we could change the default, but this would lead to other problems, for example if k3s changes its defaults or the k3s data dir is changed. I think it would be best to not generate a config for kube-proxy by default, so that k3s can decide which kubeconfig it passes to kube-proxy. However, if somebody configures a custom `extraKubeProxyConfig`, he has to take care of setting `clientConnection.kubeconfig` as mentioned in the option description.

I agree that.

File: nixos/modules/services/cluster/k3s/default.nix
Comment (rorosen) [inline nixos/modules/services/cluster/k3s/default.nix:443]: Sorry, one last thing. As users are expected to set clientConnection.kubeconfig`, it would make sense to show this in the example. Otherwise the example would be an invalid configuration.

```suggestion
      example = {
        mode = "nftables";
        clientConnection.kubeconfig = "/var/lib/rancher/k3s/agent/kubeproxy.kubeconfig";
      };
```

Comment (RMTT) [reply nixos/modules/services/cluster/k3s/default.nix:443]: oh i forgot to update example, thanks.

**Comment by rorosen** [comment]
Thanks a lot for this!

I can't resolve my own comments but they are resolved. From my side this is ready to go once you squashed the commits.

**Review by rorosen — APPROVED** [review]

**Comment by marcusramberg** [comment]
Looks great! Thanks for this.

**Pull request merged**
